proposals/gc/MVP.md

```diff
@@ -0,0 +1,304 @@
+
+* `nullref` is a subtype of `eqref`
+  - `nullref <: eqref`
+
+* `int31ref` is a subtype of `anyref`
+  - `int31ref <: anyref`
+  - Note: `int31ref` is *not* a supertype of `nullref`, i.e., nut nullable
+
+* Any nullable reference type is a subtype of `anyref` and a supertype of `nullref`
+  - `optref $t <: anyref`
+  - `nullref <: optref $t`
+
+* Any concrete reference type is a subtype of the respective nullable reference type (and thereby of `anyref`)
+  - `ref $t <: optref $t`
+  - Note: concrete reference types are *not* supertypes of `nullref`, i.e., not nullable
+
+* Any function reference type is a subtype of `anyfunc`
+  - `ref $t <: anyfunc`
+     - iff `$t = <functype>`
+
+* Any optional reference type (and thereby respective concrete reference type) is a subtype of `eqref` if its not a function
+  - `optref $t <: eqref`
+     - if `$t = <structtype>` or `$t = <arraytype>`
+     - or `$t = type rt` and `rt <: eqref`
+  - TODO: provide a way to make data types non-eq, especially immutable ones
+
+* Concrete and optional reference types are covariant
+  - `ref $t1 <: ref $t2`
+     - iff `$t1 <: $t2`
+  - `optref $t1 <: optref $t2`
+     - iff `$t1 <: $t2`
+
+* Structure types support width and depth subtyping
+  - `struct <fieldtype1>* <fieldtype1'>* <: struct <fieldtype2>*`
+    - iff `(<fieldtype1> <: <fieldtype2>)*`
+
+* Array types support depth subtyping
+  - `array <fieldtype1> <: array <fieldtype2>`
+    - iff `<fieldtype1> <: <fieldtype2>`
+
+* Field types are covariant if they are immutable, invariant otherwise
+  - `const <valtype1> <: const <valtype2>`
+    - iff `<valtype1> <: <valtype2>`
+  - `var <valtype> <: var <valtype>`
+  - Note: mutable fields are *not* subtypes of immutable ones, so `const` really means constant, not read-only
+
+
+#### Defaultability
+
+* Any numeric value type is defaultable
+
+* A reference value type is defaultable if it is not of the form `ref $t`
+
+* Locals must have a type that is defaultable.
+
+
+### Instructions
+
+#### Equality
+
+* `ref.eq` compares two references whose types support equality
+  - `ref.eq : [eqref eqref] -> [i32]`
+
+
+#### Functions
+
+* `ref.func` creates a function reference from a function index
+  - `ref.func $f : [] -> (ref $t)`
+     - iff `$f : $t`
+
+* `call_ref` calls a function through a reference
+  - `call_ref : [t1* (ref $t)] -> [t2*]`
+     - iff `$t = [t1*] -> [t2*]`
+
+
+#### Structures
+
+* `struct.new <typeidx>` allocates a structure of type `$t` and initialises its fields with given values
+  - `struct.new $t : [t*] -> [(ref $t)]`
+    - iff `$t = struct (mut t)*`
+
+* `struct.new_default <typeidx>` allocates a structure of type `$t` and initialises its fields with default values
+  - `struct.new_default $t : [] -> [(ref $t)]`
+    - iff `$t = struct (mut t)*`
+    - and all `t*` are defaultable
+
+* `struct.get <typeidx> <fieldidx>` reads field `$x` from a structure
+  - `struct.get $t i : [(ref $t)] -> [t]`
```

Comment:
For languages with ubiquitous `optref`s, `get`/`set` taking a `ref` will have a major codesize impact with all the `cast`s. It seems like `get`/`set` could take an `optref` instead (with trap semantics) and it would just be a trivial local analysis that removes the null check when the operand had static type `ref`.

Reply:
Changed. But note that this relies on subsumption, i.e., subtyping being implicit. If we were to require an explicit instruction for upcasts like we have considered recently then you would have to inversely upcast every ref to optref -- or introduce all ref instructions in two variants, or introduce overloading.